compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignNpuDmaBdIds.cpp

```diff
@@ -9,14 +9,165 @@
+            npuDmaOp.getTarget().getDefiningOp());
+    if (!logicalObjFifo)
+      return npuDmaOp.emitOpError()
+             << "expected a target `amdaie.logicalobjectfifo.from_memref`";
+    tiles = logicalObjFifo.getTiles();
+  }
+  if (tiles.size() != 1) {
+    if (tiles.empty()) {
+      return npuDmaOp.emitOpError() << "no tiles found";
+    } else {
+      return npuDmaOp.emitOpError()
+             << "operating on multiple tiles is not supported";
+    }
+  }
+  Value tile = tiles[0];
+  if (!shimTileToGeneratorMap.contains(tile))
+    return npuDmaOp.emitOpError()
+           << "no channel BD ID generator found for tile: " << tile;
+
+  auto tileOp = dyn_cast_if_present<AMDAIE::TileOp>(tile.getDefiningOp());
+  if (!tileOp) return npuDmaOp.emitOpError() << "no tile op found";
+  return tileOp;
+};
+
+// Check if the DMA operation is in the innermost loop of controlcode.
+bool isInMostInnerLoop(AMDAIE::NpuDmaCpyNdOp op) {
+  auto parentLoop = op->getParentOfType<scf::ForOp>();
+  if (!parentLoop) return false;
+
+  bool hasNestedLoop = false;
+  parentLoop.walk([&](scf::ForOp nestedLoop) {
+    if (nestedLoop != parentLoop) hasNestedLoop = true;
+  });
+  return !hasNestedLoop;
+}
+
+// Count the number of BD IDs needed per loop iteration,
+// so that we know where to start the BD ID for the next iteration.
+uint32_t countBdIdPerLoopIteration(
+    scf::ForOp loop, AMDAIE::TileOp tileOp,
+    DenseMap<Value, ChannelBdIdGenerator> &shimTileToGeneratorMap) {
+  uint32_t count = 0;
+  loop.walk([&](AMDAIE::NpuDmaCpyNdOp dmaOp) {
+    if (dmaOp.getSource() || dmaOp.getTarget()) {
+      FailureOr<AMDAIE::TileOp> tile =
+          getGeneratorTileOp(dmaOp, shimTileToGeneratorMap);
+      if (succeeded(tile) && *tile == tileOp) count++;
+    }
+  });
+  return count;
+}
+
+FailureOr<AMDAIE::BdIdOp> getBdIdOp(
+    IRRewriter &rewriter, AMDAIE::NpuDmaCpyNdOp &npuDmaOp,
+    DenseMap<Value, ChannelBdIdGenerator> &shimTileToGeneratorMap,
+    DenseMap<TileOp, uint32_t> &tileToBdIdOffsetMap,
+    DenseMap<TileOp, uint32_t> &tileToBdIdSizeMap, uint32_t channel) {
+  FailureOr<AMDAIE::TileOp> tileOp =
+      getGeneratorTileOp(npuDmaOp, shimTileToGeneratorMap);
+  if (failed(tileOp)) return failure();
+
+  ChannelBdIdGenerator &generator = shimTileToGeneratorMap[tileOp->getResult()];
+  std::optional<uint32_t> bdId =
+      generator.getAndAssignBdId(channel, BdIdAssignmentMode::Incremental);
+  if (!bdId) return failure();
+  AMDAIE::BdIdOp bdIdOp;
+  rewriter.setInsertionPoint(npuDmaOp);
+  if (isInMostInnerLoop(npuDmaOp)) {
+    // If the DMA is in the innermost loop, assign a BD ID using the
+    // semi-affine expression:
+    // `(iv * step + bdId - offset) % size + offset`,
+    // `step` represents the number of BD IDs needed per loop iteration,
+    // `bdId` is the BD ID assigned by the generator,
+    // `offset` is the BD ID assigned to the first DMA in the loop,
+    // `size` is the number of BD IDs available.
```

Comment:
I don't think I am fully understanding this expression. 

Assuming the loop is normalized (we can normalize and ensure this before), I would have expected:
```
size = max(<available_incremental_bd_ids>, <loop_extent>)
bd_id = offset + (iv % size)
```

Reply:
that's true if there is only one DMA copy inside the loop, where in my expression `step = 1` and `bdId-offset=0`. Otherwise, in a different example `step=2`, `offset=1`, `size=16-1=15`

```ll
dma_cpy_nd(bd_id = 0) // generator assign bd_id 0, so `offset=1`
scf.for iv {
   dma_cpy_nd(bd_id = (2 * iv + 1 - 1) % 15 + 1) // generator assign bd_id 1 for the first iteration, then it is converted to expression
   dma_cpy_nd(bd_id = (2 * iv + 2 - 1) % 15 + 1)   // generator assign bd_id 2 for the first iteration, then it is converted to expression
}
```

and I can't move `+-offset` inside or outside the bracket due to the modulo operation

Reply:
Ok, I see, that's more general towards multiple `dma_cpy_nd`, cool!

A few follow-up question: where do you ensure that all these BDs being used in the loop are actually assigned?